File: Data Structures & Algorithms/count-connected-components/submission-0.cpp
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> graph(n);
        for (auto& e : edges) {
            graph[e[0]].push_back(e[1]);
            graph[e[1]].push_back(e[0]);
        }

        vector<bool> visited(n, false);

        function <void(int, int)> dfs = [&] (int node, int parent) -> void {
            visited[node] = true;

            for (int nei : graph[node]) {
                if (visited[nei]) continue;
                if (parent == nei) continue;
                dfs(nei, node);
            }
            return;
        };

        int connected = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                dfs(i, -1);
                connected++;
            }
        }

        return connected;
    }
};

File: Data Structures & Algorithms/non-cyclical-number/submission-0.py
class Solution:
    def isHappy(self, n: int) -> bool:
        visited = set()
        j = 0
        while True:
            if n == 1:
                return True
            
            if n in visited:
                return False
            
            visited.add(n)

            x = 0
            while n > 0:
                x += (n % 10) ** 2
                n = n // 10
            n = x
            # print(n)
            # j += 1
            # if j == 10:
            #     break
            # print(n)

        return False